src/vsock/src/vsock_conn.rs

```diff
@@ -0,0 +1,329 @@
+            // Data has to be written to the host-side stream
+            if pkt.buf().is_none() {
+                info!(
+                    "Dropping empty packet from guest (lp={}, pp={})",
+                    self.local_port, self.peer_port
+                );
+                return Ok(());
+            }
+
+            let buf_slice = &pkt.buf().unwrap()[..(pkt.len() as usize)];
+            if let Err(err) = self.send_bytes(buf_slice) {
+                // TODO: Terminate this connection
+                dbg!("err:{:?}", err);
+                return Ok(());
+            }
+        } else if pkt.op() == VSOCK_OP_CREDIT_UPDATE {
+            // Already updated the credit
+
+            // Re-register the stream file descriptor for read and write events
+            if VhostUserVsockThread::epoll_modify(
+                self.epoll_fd,
+                self.stream.as_raw_fd(),
+                epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+            )
+            .is_err()
+            {
+                VhostUserVsockThread::epoll_register(
+                    self.epoll_fd,
+                    self.stream.as_raw_fd(),
+                    epoll::Events::EPOLLIN | epoll::Events::EPOLLOUT,
+                )
+                .unwrap();
+            };
+        } else if pkt.op() == VSOCK_OP_CREDIT_REQUEST {
+            // Send back this connection's credit information
+            self.rx_queue.enqueue(RxOps::CreditUpdate);
+        } else if pkt.op() == VSOCK_OP_SHUTDOWN {
+            // Shutdown this connection
+            let recv_off = pkt.flags() & VSOCK_FLAGS_SHUTDOWN_RCV != 0;
+            let send_off = pkt.flags() & VSOCK_FLAGS_SHUTDOWN_SEND != 0;
+
+            if recv_off && send_off && self.tx_buf.is_empty() {
+                self.rx_queue.enqueue(RxOps::Reset);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Write data to the host-side stream.
+    ///
+    /// Returns:
+    /// - Ok(cnt) where cnt is the number of bytes written to the stream
+    /// - Err(Error::UnixWrite) if there was an error writing to the stream
+    fn send_bytes(&mut self, buf: &[u8]) -> Result<()> {
+        if !self.tx_buf.is_empty() {
+            // Data is already present in the buffer and the backend
+            // is waiting for a EPOLLOUT event to flush it
+            return self.tx_buf.push(buf);
+        }
+
+        // Write data to the stream
+        let written_count = match self.stream.write(buf) {
+            Ok(cnt) => cnt,
+            Err(e) => {
+                if e.kind() == ErrorKind::WouldBlock {
+                    0
+                } else {
+                    println!("send_bytes error: {:?}", e);
+                    return Err(Error::UnixWrite);
+                }
+            }
+        };
+
+        // Increment forwarded count by number of bytes written to the stream
+        self.fwd_cnt += Wrapping(written_count as u32);
+        self.rx_queue.enqueue(RxOps::CreditUpdate);
```

Comment:
An improvement we did in `vhost-vsock` kernel module to increase the performance was to reduce the number of credit messages. We avoid to send them every time we consume bytes, but only after a threshold.

More details here: https://github.com/torvalds/linux/commit/c69e6eafff5f725bc29dcb8b52b6782dca8ea8a2

This could be another TODO idem, not to resolve now :-)

Reply:
I need some time to test this out, so just added a TODO for now.

Reply:
Yep sure, I just wanted to let you know, there's no need to do it now, it's just an optimization :-)